File: LeetCode/454四数相加Ⅱ.cpp
class Solution {
public:
    int fourSumCount(vector<int>& nums1, vector<int>& nums2, vector<int>& nums3, vector<int>& nums4) {
        int ans = 0;
        unordered_map<int, int> u_map;
        for (int a : nums1) {
            for (int b : nums2) {
                u_map[a + b]++;
            }
        }
        for (int a : nums3) {
            for (int b : nums4) {
                int sum = 0 - a - b;
                if (u_map[sum] != 0) {
                    ans += u_map[sum];
                }
            }
        }
        return ans;
    }
};

File: LeetCode/454四数相加Ⅱ.go
func fourSumCount(nums1 []int, nums2 []int, nums3 []int, nums4 []int) int {
	set := map[int]int{}
	ans := 0
	for _, v1 := range nums1 {
		for _, v2 := range nums2 {
			sum := v1 + v2
			if set[sum] != 0 {
				set[sum] += 1
			} else {
				set[sum] = 1
			}
		}
	}
	for _, v1 := range nums3 {
		for _, v2 := range nums4 {
			sum := 0 - v1 - v2
			if set[sum] != 0 {
				ans += set[sum]
			}
		}
	}
	return ans
}